Add structure to the code

arastoob opened the pull request:
This will help us have a structural code. The main boxes are organized in their own structures and later we can add more functionalities to them, e.g., add some functions to the `Movie` structure to get the corresponding video frame of a specific time period.

arastoob committed:
Ignore more files

arastoob committed:
Use the main name for the file as it contains the main function

arastoob committed:
Use StructOps to handle the inputs

arastoob committed:
Handle the library errors

arastoob committed:
Remove extra imports

arastoob committed:
Implement Mp4Atom for different atom types

arastoob committed:
Handle Utf8 errors

arastoob committed:
Implement the required debug and display

arastoob committed:
Pass the correct indices

arastoob committed:
Create the mp4 structure and print it

arastoob committed:
Get the internal structure of an atom if exist

arastoob committed:
Fix the copyright date

arastoob committed:
Ignore the sample data directory

arastoob committed:
Add a header to the structs and use it

arastoob committed:
Disable the previous code for now

arastoob committed:
Use a better naming for the modules and re-export the required modules

arastoob committed:
Fix some of the compiler warnings

arastoob committed:
Resolve more warnings

arastoob committed:
Remove unnecessary comments

arastoob committed:
Add a level field to help the output messages

arastoob committed:
Some minor changes to the output

arastoob committed:
Add some comments

arastoob committed:
Use better and more meaningful names

arastoob committed:
Fix the readme instructions

arastoob committed:
Rename Ftype to FileType

arastoob committed:
Rename Udata to UserData

arastoob committed:
Handle missed boxes

arastoob committed:
Add boxes as struct fields and their accessor methods

arastoob committed:
Apply auto-format

sdharar commented on src/header.rs:
```diff
@@ -0,0 +1,25 @@
+}
+
+impl Header {
+    pub fn header(data: &[u8], start: usize) -> Result<Self, Error> {
```
warning: constructor `header` has the same name as the type
  --> src\header.rs:17:5
   |
17 | /     pub fn header(data: &[u8], start: usize) -> Result<Self, Error> {
18 | |         // the first 8 bytes includes the atom size and its name
19 | |         // The size is the entire size of the box, including the size and type header, fields, and all contained boxes.
20 | |         let size = BigEndian::read_u32(&data[0..4]) as usize;
...  |
23 | |         Ok(Self { name, size, start })
24 | |     }
   | |_____^
   |
   = note: `#[warn(clippy::self_named_constructors)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#self_named_constructors

sdharar commented on src/boxes/file_type.rs:
```diff
@@ -0,0 +1,54 @@
+#[derive(Debug)]
+pub struct FileType {
+    header: Header,
+    data: Vec<u8>,
```
warning: field is never read: `data`
  --> src\boxes\file_type.rs:13:5
   |
13 |     data: Vec<u8>,
   |     ^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

sdharar commented on src/boxes/media_data.rs:
```diff
@@ -0,0 +1,54 @@
+#[derive(Debug)]
+pub struct MediaData {
+    header: Header,
+    data: Vec<u8>,
```
warning: field is never read: `data`
  --> src\boxes\media_data.rs:13:5
   |
13 |     data: Vec<u8>,

sdharar commented on src/boxes/mp4_box.rs:
```diff
@@ -0,0 +1,127 @@
+    fn read(&self) -> Result<Vec<u8>, Error>;
+
+    /// Get the internal boxes of this box
+    fn fields(&self) -> Option<Vec<&Box<dyn Mp4Box>>>;
```
warning: you seem to be trying to use `&Box<T>`. Consider using just `&T`
  --> src\boxes\mp4_box.rs:31:36
   |
31 |     fn fields(&self) -> Option<Vec<&Box<dyn Mp4Box>>>;
   |                                    ^^^^^^^^^^^^^^^^ help: try: `&dyn Mp4Box`
   |
   = note: `#[warn(clippy::borrowed_box)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrowed_box

sdharar commented on src/boxes/mp4_box.rs:
```diff
@@ -0,0 +1,127 @@
+        )?;
+
+        let internals = self.fields();
+        if internals.is_some() {
```
warning: called `unwrap` on `internals` after checking its variant with `is_some`
  --> src\boxes\mp4_box.rs:51:29
   |
50 |         if internals.is_some() {
   |         ---------------------- help: try: `if let Some(..) = internals`
51 |             for internal in internals.unwrap() {
   |                             ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::unnecessary_unwrap)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap

sdharar commented on src/bin/main.rs:
```diff
@@ -0,0 +1,29 @@
+)]
+struct Opt {
+    #[structopt(short, long)]
+    verbose: bool,
```
warning: field is never read: `verbose`
  --> src\bin\main.rs:16:5
   |
16 |     verbose: bool,
   |     ^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

sdharar reviewed (changes requested):
Dr. @arastoob => The structure of the code has improved a lot, and it appears to generate the expected results. Upon running some tests, I have stumbled upon some changes that have been added as comments.

If you think the code requires changes based on my comments, perhaps those changes can be made before committing this PR.

arastoob committed:
Rename the Header constructor

arastoob committed:
Return the box data

arastoob committed:
Use if let instead of is_some and unwrap

arastoob committed:
Apply more clippy suggestions

arastoob commented:
> 
@sdharar  What comments do you mean? If you mean `reader` and `boxes_old` files, they are your file and I thought we might need it. If you think that the new code has covered what is inside those files, we can remove them?

sdharar commented:
> 
No, I meant the comments that I added on your code. Yes, the new code does pretty much outputs the same thing that older code did, in addition to one more layer further down in track/moov atoms but its more structured. The ftype atom needs more further layer to be implemented. Perhaps we can take some implementations from the older code for ftype atom. What do you think?

arastoob commented:
> > 
> 
> No, I meant the comments that I added on your code. Yes, the new code does pretty much outputs the same thing that older code did, in addition to one more layer further down in track/moov atoms but its more structured. The ftype atom needs more further layer to be implemented. Perhaps we can take some implementations from the older code for ftype atom. What do you think?

So where can I see the comments?
and yes, please go ahead and add your code to ftype.

sdharar commented:
> > > 
> > 
> > 
> > No, I meant the comments that I added on your code. Yes, the new code does pretty much outputs the same thing that older code did, in addition to one more layer further down in track/moov atoms but its more structured. The ftype atom needs more further layer to be implemented. Perhaps we can take some implementations from the older code for ftype atom. What do you think?
> 
> So where can I see the comments? and yes, please go ahead and add your code to ftype.

I added suggestions I received from clippy on this webpage, which I guess you already have rectified.

arastoob commented:
> > > > 
> > > 
> > > 
> > > No, I meant the comments that I added on your code. Yes, the new code does pretty much outputs the same thing that older code did, in addition to one more layer further down in track/moov atoms but its more structured. The ftype atom needs more further layer to be implemented. Perhaps we can take some implementations from the older code for ftype atom. What do you think?
> > 
> > 
> > So where can I see the comments? and yes, please go ahead and add your code to ftype.
> 
> I added suggestions I received from clippy on this webpage, which I guess you already have rectified.

Yes, I have resolved those comments. Sorry, I thought that you have made some comments inside the code. So, I can merge the code into master.

arastoob committed:
Remove the commented codes
